### PR TITLE
Introduce `as_components` interface for Python APIs

### PR DESCRIPTION
**Pull request by jleibs**
### What
`rr2.log` now makes use of two newly defined protocols:
```python
class ComponentBatchLike(Protocol):
    """Describes interface for objects that can be converted to rerun Components."""

    def component_name(self) -> str:
        """Returns the name of the component."""
        ...

    def as_arrow_batch(self) -> pa.Array:
        """
        Returns a `pyarrow.Array` of the component data.

        Each element in the array corresponds to an instance of the component. Single-instanced
        components and splats must still be represented as a 1-element array.
        """
        ...


class ArchetypeLike(Protocol):
    """Describes interface for objects that can be logged via rr.log."""

    def as_component_batches(self) -> Iterable[ComponentBatchLike]:
        """
        Returns an iterable of `ComponentBatchLike` objects.

        Each object in the iterable must adhere to the `ComponentBatchLike`
        interface. All of the batches should have the same length as the value
        returned by `num_instances`, or length 1 if the component is a splat.,
        or 0 if the component is being cleared.
        """
        ...

    def num_instances(self) -> int:
        """
        The number of instances in each batch.

        Each batch returned by `as_component_batches` should have this number of
        elements, or 1 in the case it is a splat, or 0 in the case that
        component is being cleared.
        """
        ...
```

Additionally, the python `Archetype` and `BaseExtensionArray` have been updated to adhere to these protocols.

### Notable Changes

This moves the decision of whether or not to log empty lists from the logger to the archetypes implementation of `as_components` so that the logger code can deterministically pass through any components returned by the archetype.

I introduced a new `optional_from_similar` converter that we use for optional components so we can more easily track which fields were explicitly set or not.

I also created https://github.com/rerun-io/rerun/issues/3341 to track/discuss whether these should be turned into empty lists.

### In Action
```python
from __future__ import annotations

import numpy as np
import pyarrow as pa
import rerun as rr
import rerun.experimental as rr2

rr.init("my-points", spawn=True)


class ConfidenceBatch:
    def __init__(self, confidence: np.ArrayLike) -> None:
        self.data = confidence

    def as_arrow_batch(self) -> pa.Array:
        return pa.array(self.data, type=pa.float32())

    def component_name(self) -> str:
        return "Confidence"


class MyPointsWithConfidence:
    def __init__(self, points: np.ArrayLike, confidence: np.ArrayLike) -> None:
        self.points = points
        self.confidence = confidence

    def as_component_batches(self) -> list[rr2.ComponentBatchLike]:
        return list(rr2.Points3D(self.points).as_component_batches()) + [ConfidenceBatch(self.confidence)]

    def num_instances(self) -> int:
        return len(self.points)


my_points = MyPointsWithConfidence([[0, 0, 0], [1, 1, 1]], confidence=[0.5, 0.6])

rr2.log("points", my_points)
```

image

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested [demo.rerun.io](https://demo.rerun.io/pr/3340) (if applicable)

- [PR Build Summary](https://build.rerun.io/pr/3340)
- [Docs preview](https://rerun.io/preview/e3bff85cffd1552b0fb261984430ea9c158567de/docs) 
- [Examples preview](https://rerun.io/preview/e3bff85cffd1552b0fb261984430ea9c158567de/examples) 
- [Recent benchmark results](https://ref.rerun.io/dev/bench/)
- [Wasm size tracking](https://ref.rerun.io/dev/sizes/)

### CODE REVIEW

File: rerun_py/rerun_sdk/rerun/_rerun2/log.py [hunk 1]
```diff
@@ -22,6 +19,50 @@
 ext_component_types: dict[str, Any] = {}
 
 
+@runtime_checkable
+class ComponentBatchLike(Protocol):
+    """Describes interface for objects that can be converted to rerun Components."""
```
Comment (teh-cmc) [inline rerun_py/rerun_sdk/rerun/_rerun2/log.py:24]:
```suggestion
    """Describes interface for objects that can be converted to batches of Rerun components."""

**Review by teh-cmc — APPROVED** [review]
Looks great! We should implement the `custom_data` example for Python before merging though, we should have everything we need to at this point?

Also we might want to move `custom_data` to API examples once we do, so we benefit from roundtrips on this one too.

**Comment by jleibs** [comment]
Resolves: https://github.com/rerun-io/rerun/issues/3254, https://github.com/rerun-io/rerun/issues/3253,

**Comment by teh-cmc** [comment]
So nice to see some crosslang roundtripped custom data :scream: thanks!

**Pull request merged**